File: typescript/src/transport/http.ts
import type { VectorlessConfig } from "../config.js";
import type { Transport } from "./interface.js";
import type {
  HealthResponse,
  VersionResponse,
  Document,
  IngestDocumentOptions,
  IngestDocumentResponse,
  ListDocumentsOptions,
  ListDocumentsResponse,
  DocumentTree,
  Section,
  QueryOptions,
  QueryResponse,
  QueryStreamEvent,
} from "../types.js";
import { withRetry } from "../retry.js";
import { prepareUpload } from "../upload.js";
import { parseSSEStream } from "../streaming.js";
import {
  VectorlessError,
  AuthenticationError,
  PermissionDeniedError,
  NotFoundError,
  RateLimitError,
  ValidationError,
  ConflictError,
  ServerError,
  TimeoutError,
} from "../errors.js";

const SDK_VERSION = "1.0.0";

/**
 * HTTP/REST transport.
 *
 * Talks to the Vectorless server via standard REST endpoints:
 *   GET/POST/DELETE /v1/documents, /v1/sections, /v1/query, etc.
 *
 * Streaming queries use Server-Sent Events (SSE) via POST /v1/query/stream.
 *
 * This is the default transport — zero extra dependencies beyond `fetch`.
 */
export class HttpTransport implements Transport {
  private readonly config: VectorlessConfig;

  constructor(config: VectorlessConfig) {
    this.config = config;
  }

  // ── Health ──

  async health(): Promise<HealthResponse> {
    return this.request<HealthResponse>("GET", "/v1/health");
  }

  async version(): Promise<VersionResponse> {
    return this.request<VersionResponse>("GET", "/v1/version");
  }

  // ── Documents ──

  async ingestDocument(
    source: Buffer | ArrayBuffer | Blob | string,
    options?: IngestDocumentOptions
  ): Promise<IngestDocumentResponse> {
    const formData = await prepareUpload(source, options);

    const headers: Record<string, string> = {};
    if (options?.idempotencyKey) {
      headers["Idempotency-Key"] = options.idempotencyKey;
    }

    return this.request<IngestDocumentResponse>(
      "POST",
      "/v1/documents",
      formData,
      { isMultipart: true, extraHeaders: headers }
    );
  }

  async getDocument(documentId: string): Promise<Document> {
    return this.request<Document>("GET", `/v1/documents/${documentId}`);
  }

  async listDocuments(
    options?: ListDocumentsOptions
  ): Promise<ListDocumentsResponse> {
    const params = new URLSearchParams();
    if (options?.limit != null) params.set("limit", String(options.limit));
    if (options?.cursor) params.set("cursor", options.cursor);
    if (options?.status) params.set("status", options.status);

    const qs = params.toString();
    const path = qs ? `/v1/documents?${qs}` : "/v1/documents";
    return this.request<ListDocumentsResponse>("GET", path);
  }

  async deleteDocument(documentId: string): Promise<void> {
    await this.request<void>("DELETE", `/v1/documents/${documentId}`);
  }

  async getDocumentTree(documentId: string): Promise<DocumentTree> {
    return this.request<DocumentTree>(
      "GET",
      `/v1/documents/${documentId}/tree`
    );
  }

  // ── Sections ──

  async getSection(sectionId: string): Promise<Section> {
    return this.request<Section>("GET", `/v1/sections/${sectionId}`);
  }

  // ── Query ──

  async query(
    documentId: string,
    query: string,
    options?: QueryOptions
  ): Promise<QueryResponse> {
    return this.request<QueryResponse>("POST", "/v1/query", {
      document_id: documentId,
      query,
      model: options?.model,
      max_tokens: options?.maxTokens,
      reserved_for_prompt: options?.reservedForPrompt,
      max_parallel_calls: options?.maxParallelCalls,
      max_sections: options?.maxSections,
    });
  }

  async *queryStream(
    documentId: string,
    query: string,
    options?: QueryOptions
  ): AsyncIterable<QueryStreamEvent> {
    const url = `${this.config.baseUrl}/v1/query/stream`;
    const headers = this.buildHeaders();
    headers["Content-Type"] = "application/json";
    headers["Accept"] = "text/event-stream";

    const body = JSON.stringify({
      document_id: documentId,
      query,
      model: options?.model,
      max_tokens: options?.maxTokens,
      reserved_for_prompt: options?.reservedForPrompt,
      max_parallel_calls: options?.maxParallelCalls,
      max_sections: options?.maxSections,
    });

    const response = await fetch(url, {
      method: "POST",
      headers,
      body,
    });

    if (!response.ok) {
      throw await this.parseErrorResponse(response);
    }

    yield* parseSSEStream(response);
  }

  // ── Lifecycle ──

  async close(): Promise<void> {
    // No persistent connections with fetch — nothing to close
  }

  // ── Internal ──

  private buildHeaders(): Record<string, string> {
    const headers: Record<string, string> = {
      "User-Agent": `vectorless-ts/${SDK_VERSION}`,
    };
    if (this.config.apiKey) {
      headers["Authorization"] = `Bearer ${this.config.apiKey}`;
    }
    if (this.config.store) {
      headers["X-Vectorless-Store"] = this.config.store;
    }
    return headers;
  }

  private async request<T>(
    method: string,
    path: string,
    body?: unknown,
    opts?: { isMultipart?: boolean; extraHeaders?: Record<string, string> }
  ): Promise<T> {
    const url = `${this.config.baseUrl}${path}`;
    const headers = this.buildHeaders();

    if (opts?.extraHeaders) {
      Object.assign(headers, opts.extraHeaders);
    }

    if (!opts?.isMultipart && body && !(body instanceof FormData)) {
      headers["Content-Type"] = "application/json";
    }

    const fetchBody =
      body instanceof FormData
        ? body
        : body
          ? JSON.stringify(body)
          : undefined;

    return withRetry(
      async () => {
        const controller = new AbortController();
        const timeoutId = setTimeout(
          () => controller.abort(),
          this.config.timeout
        );

        try {
          const response = await fetch(url, {
            method,
            headers,
            body: fetchBody,
            signal: controller.signal,
          });

          if (!response.ok) {
            throw await this.parseErrorResponse(response);
          }

          if (response.status === 204) return undefined as T;
          return (await response.json()) as T;
        } catch (error) {
          if (error instanceof VectorlessError) throw error;
          if (error instanceof DOMException && error.name === "AbortError") {
            throw new TimeoutError(
              `Request timed out after ${this.config.timeout}ms`
            );
          }
          throw error;
        } finally {
          clearTimeout(timeoutId);
        }
      },
      this.config.maxRetries,
      this.config.retryDelay
    );
  }

  private async parseErrorResponse(
    response: Response
  ): Promise<VectorlessError> {
    const requestId = response.headers.get("x-request-id") ?? undefined;

    try {
      const body = (await response.json()) as {
        error?: string | {
          code?: string;
          message?: string;
          field_errors?: Record<string, string[]>;
        };
      };

      const message =
        typeof body.error === "string"
          ? body.error
          : body.error?.message ?? `Request failed with status ${response.status}`;

      const fieldErrors =
        typeof body.error === "object" ? body.error?.field_errors : undefined;

      switch (response.status) {
        case 400:
          return new ValidationError(message, fieldErrors, requestId);
        case 401:
          return new AuthenticationError(message, requestId);
        case 403:
          return new PermissionDeniedError(message, requestId);
        case 404:
          return new NotFoundError(message, requestId);
        case 409:
          return new ConflictError(message, requestId);
        case 429: {
          const retryAfter = response.headers.get("retry-after");
          return new RateLimitError(
            message,
            retryAfter ? parseInt(retryAfter, 10) : undefined,
            requestId
          );
        }
        case 408:
        case 504:
          return new TimeoutError(message, requestId);
        default:
          if (response.status >= 500) {
            return new ServerError(message, requestId);
          }
          return new VectorlessError(
            message,
            response.status,
            typeof body.error === "object"
              ? body.error?.code ?? "unknown"
              : "unknown",
            requestId
          );
      }
    } catch {
      return new VectorlessError(
        `Request failed with status ${response.status}`,
        response.status,
        "unknown",
        requestId
      );
    }
  }
}

File: go/transport/http.go
package transport

import (
	"bufio"
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"iter"
	"mime/multipart"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"time"

	vectorless "github.com/hallelx2/vectorless-sdk/go"
)

const sdkVersion = "1.0.0"

// HTTPTransport implements Transport over standard REST/JSON endpoints.
type HTTPTransport struct {
	client  *http.Client
	baseURL string
	apiKey  string
	store   string
}

// NewHTTPTransport creates a new HTTP/REST transport.
func NewHTTPTransport(cfg vectorless.Config) *HTTPTransport {
	return &HTTPTransport{
		client: &http.Client{
			Timeout: time.Duration(cfg.TimeoutMs) * time.Millisecond,
		},
		baseURL: strings.TrimRight(cfg.BaseURL, "/"),
		apiKey:  cfg.APIKey,
		store:   cfg.Store,
	}
}

// ── Health ──

func (t *HTTPTransport) Health(ctx context.Context) (*vectorless.HealthResponse, error) {
	var out vectorless.HealthResponse
	if err := t.get(ctx, "/v1/health", nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

func (t *HTTPTransport) Version(ctx context.Context) (*vectorless.VersionResponse, error) {
	var out vectorless.VersionResponse
	if err := t.get(ctx, "/v1/version", nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// ── Documents ──

func (t *HTTPTransport) IngestDocument(ctx context.Context, body io.Reader, opts vectorless.IngestDocumentOptions) (*vectorless.IngestDocumentResponse, error) {
	var buf bytes.Buffer
	w := multipart.NewWriter(&buf)

	filename := opts.Filename
	if filename == "" {
		filename = "document"
	}

	part, err := w.CreateFormFile("file", filename)
	if err != nil {
		return nil, fmt.Errorf("vectorless: create form file: %w", err)
	}
	if _, err := io.Copy(part, body); err != nil {
		return nil, fmt.Errorf("vectorless: copy body: %w", err)
	}

	if opts.ContentType != "" {
		_ = w.WriteField("content_type", opts.ContentType)
	}
	for k, v := range opts.Metadata {
		_ = w.WriteField(fmt.Sprintf("metadata[%s]", k), v)
	}
	w.Close()

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, t.baseURL+"/v1/documents", &buf)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", w.FormDataContentType())
	t.setHeaders(req)

	if opts.IdempotencyKey != "" {
		req.Header.Set("Idempotency-Key", opts.IdempotencyKey)
	}

	var out vectorless.IngestDocumentResponse
	if err := t.doJSON(req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

func (t *HTTPTransport) GetDocument(ctx context.Context, documentID string) (*vectorless.Document, error) {
	var out vectorless.Document
	if err := t.get(ctx, "/v1/documents/"+documentID, nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

func (t *HTTPTransport) ListDocuments(ctx context.Context, opts vectorless.ListDocumentsOptions) (*vectorless.ListDocumentsResponse, error) {
	params := url.Values{}
	if opts.Limit > 0 {
		params.Set("limit", strconv.Itoa(opts.Limit))
	}
	if opts.Cursor != "" {
		params.Set("cursor", opts.Cursor)
	}
	if opts.Status != "" {
		params.Set("status", string(opts.Status))
	}

	var out vectorless.ListDocumentsResponse
	if err := t.get(ctx, "/v1/documents", params, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

func (t *HTTPTransport) DeleteDocument(ctx context.Context, documentID string) error {
	req, err := http.NewRequestWithContext(ctx, http.MethodDelete, t.baseURL+"/v1/documents/"+documentID, nil)
	if err != nil {
		return err
	}
	t.setHeaders(req)
	return t.doJSON(req, nil)
}

func (t *HTTPTransport) GetDocumentTree(ctx context.Context, documentID string) (*vectorless.DocumentTree, error) {
	var out vectorless.DocumentTree
	if err := t.get(ctx, "/v1/documents/"+documentID+"/tree", nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// ── Sections ──

func (t *HTTPTransport) GetSection(ctx context.Context, sectionID string) (*vectorless.Section, error) {
	var out vectorless.Section
	if err := t.get(ctx, "/v1/sections/"+sectionID, nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// ── Query ──

func (t *HTTPTransport) Query(ctx context.Context, documentID, query string, opts *vectorless.QueryOptions) (*vectorless.QueryResponse, error) {
	body := buildQueryBody(documentID, query, opts)

	var out vectorless.QueryResponse
	if err := t.postJSON(ctx, "/v1/query", body, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

func (t *HTTPTransport) QueryStream(ctx context.Context, documentID, query string, opts *vectorless.QueryOptions) (iter.Seq2[vectorless.QueryStreamEvent, error], error) {
	body := buildQueryBody(documentID, query, opts)

	jsonBody, err := json.Marshal(body)
	if err != nil {
		return nil, err
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, t.baseURL+"/v1/query/stream", bytes.NewReader(jsonBody))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Accept", "text/event-stream")
	t.setHeaders(req)

	resp, err := t.client.Do(req)
	if err != nil {
		return nil, err
	}

	if resp.StatusCode >= 400 {
		defer resp.Body.Close()
		return nil, parseHTTPError(resp)
	}

	return func(yield func(vectorless.QueryStreamEvent, error) bool) {
		defer resp.Body.Close()
		scanner := bufio.NewScanner(resp.Body)

		var eventType string
		for scanner.Scan() {
			line := scanner.Text()
			if strings.HasPrefix(line, "event:") {
				eventType = strings.TrimSpace(line[6:])
			} else if strings.HasPrefix(line, "data:") {
				data := strings.TrimSpace(line[5:])
				var event vectorless.QueryStreamEvent
				if err := json.Unmarshal([]byte(data), &event); err != nil {
					if !yield(vectorless.QueryStreamEvent{}, err) {
						return
					}
					continue
				}
				if eventType != "" && event.Type == "" {
					event.Type = vectorless.QueryStreamEventType(eventType)
				}
				eventType = ""
				if !yield(event, nil) {
					return
				}
			}
		}
		if err := scanner.Err(); err != nil {
			yield(vectorless.QueryStreamEvent{}, err)
		}
	}, nil
}

// ── Lifecycle ──

func (t *HTTPTransport) Close() error {
	t.client.CloseIdleConnections()
	return nil
}

// ── Internal ──

func (t *HTTPTransport) setHeaders(req *http.Request) {
	req.Header.Set("User-Agent", "vectorless-go/"+sdkVersion)
	if t.apiKey != "" {
		req.Header.Set("Authorization", "Bearer "+t.apiKey)
	}
	if t.store != "" {
		req.Header.Set("X-Vectorless-Store", t.store)
	}
}

func (t *HTTPTransport) get(ctx context.Context, path string, params url.Values, out any) error {
	u := t.baseURL + path
	if len(params) > 0 {
		u += "?" + params.Encode()
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, u, nil)
	if err != nil {
		return err
	}
	t.setHeaders(req)
	return t.doJSON(req, out)
}

func (t *HTTPTransport) postJSON(ctx context.Context, path string, body any, out any) error {
	jsonBody, err := json.Marshal(body)
	if err != nil {
		return err
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, t.baseURL+path, bytes.NewReader(jsonBody))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")
	t.setHeaders(req)
	return t.doJSON(req, out)
}

func (t *HTTPTransport) doJSON(req *http.Request, out any) error {
	resp, err := t.client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode >= 400 {
		return parseHTTPError(resp)
	}

	if resp.StatusCode == 204 || out == nil {
		return nil
	}

	return json.NewDecoder(resp.Body).Decode(out)
}

func parseHTTPError(resp *http.Response) *vectorless.Error {
	requestID := resp.Header.Get("X-Request-ID")

	var body struct {
		Error any `json:"error"`
	}
	message := fmt.Sprintf("Request failed with status %d", resp.StatusCode)

	if err := json.NewDecoder(resp.Body).Decode(&body); err == nil {
		switch v := body.Error.(type) {
		case string:
			message = v
		case map[string]any:
			if m, ok := v["message"].(string); ok {
				message = m
			}
		}
	}

	switch resp.StatusCode {
	case 400:
		return &vectorless.Error{Message: message, Status: 400, Code: "validation_error", RequestID: requestID}
	case 401:
		return &vectorless.Error{Message: message, Status: 401, Code: "authentication_error", RequestID: requestID}
	case 403:
		return &vectorless.Error{Message: message, Status: 403, Code: "permission_denied", RequestID: requestID}
	case 404:
		return &vectorless.Error{Message: message, Status: 404, Code: "not_found", RequestID: requestID}
	case 409:
		return &vectorless.Error{Message: message, Status: 409, Code: "conflict", RequestID: requestID}
	case 429:
		return &vectorless.Error{Message: message, Status: 429, Code: "rate_limit_exceeded", RequestID: requestID}
	case 408, 504:
		return &vectorless.Error{Message: message, Status: resp.StatusCode, Code: "timeout", RequestID: requestID}
	default:
		if resp.StatusCode >= 500 {
			return &vectorless.Error{Message: message, Status: resp.StatusCode, Code: "server_error", RequestID: requestID}
		}
		return &vectorless.Error{Message: message, Status: resp.StatusCode, Code: "unknown", RequestID: requestID}
	}
}

func buildQueryBody(documentID, query string, opts *vectorless.QueryOptions) map[string]any {
	body := map[string]any{
		"document_id": documentID,
		"query":       query,
	}
	if opts != nil {
		if opts.Model != "" {
			body["model"] = opts.Model
		}
		if opts.MaxTokens > 0 {
			body["max_tokens"] = opts.MaxTokens
		}
		if opts.ReservedForPrompt > 0 {
			body["reserved_for_prompt"] = opts.ReservedForPrompt
		}
		if opts.MaxParallelCalls > 0 {
			body["max_parallel_calls"] = opts.MaxParallelCalls
		}
		if opts.MaxSections > 0 {
			body["max_sections"] = opts.MaxSections
		}
	}
	return body
}
